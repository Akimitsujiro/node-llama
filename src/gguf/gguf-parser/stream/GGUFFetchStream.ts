import retry from "async-retry";
import AwaitLock from "await-lock";
import GgufBaseStream, {ALLOCATION_SIZE} from "./GGUFBaseStream.js";

type GGUFFetchStreamOptions = {
    retry: retry.Options
};

export default class GGUFFetchStream extends GgufBaseStream {
    private _lock = new ("default" in AwaitLock && AwaitLock || AwaitLock)();

    constructor(public readonly url: string, public readonly options: Partial<GGUFFetchStreamOptions> = {}) {
        super();
    }

    override async readNBytes(numBytes: number, offset = 0): Promise<Buffer> {
        await this._lock.acquireAsync();

        try {
            if (offset + numBytes < this._buffer.length) {
                return this._buffer.subarray(offset, offset + numBytes);
            }

            const fetchMissingBytes = await retry(async () => {
                return await this._fetchBytesWithoutRetry(this._buffer.length, offset + numBytes + ALLOCATION_SIZE);
            }, this.options.retry);

            this._addToBuffer(fetchMissingBytes);
            return this.readNBytes(numBytes, offset);
        } finally {
            this._lock.release();
        }
    }

    private async _fetchBytesWithoutRetry(start: number, end: number) {
        const response = await fetch(this.url, {
            headers: {
                Range: `bytes=${start}-${end}`,
                accept: "*/*"
            }
        });
        const arrayBuffer = await response.arrayBuffer();
        return Buffer.from(arrayBuffer);
    }
}


