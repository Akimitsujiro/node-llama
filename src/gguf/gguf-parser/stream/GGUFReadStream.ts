import fs from "fs/promises";
import retry from "async-retry";
import AwaitLock from "await-lock";
import GgufBaseStream, {ALLOCATION_SIZE} from "./GGUFBaseStream.js";

type GGUFReadStreamOptions = {
    retry?: retry.Options
    mode: string;
};

const DEFAULT_OPTIONS: GGUFReadStreamOptions = {
    mode: "r"
};

export default class GGUFReadStream extends GgufBaseStream {
    private _lock = new ("default" in AwaitLock && AwaitLock || AwaitLock)();
    public readonly options: GGUFReadStreamOptions;

    constructor(public readonly path: string, options: Partial<GGUFReadStreamOptions> = {}) {
        super();
        this.options = {...DEFAULT_OPTIONS, ...options};
    }

    override async readNBytes(numBytes: number, offset = 0): Promise<Buffer> {
        await this._lock.acquireAsync();

        try {
            if (offset + numBytes < this._buffer.length) {
                return this._buffer.subarray(offset, offset + numBytes);
            }

            const readMissingBytes = await retry(async () => {
                return await this._readBytesWithoutRetry(numBytes + ALLOCATION_SIZE, this._buffer.length);
            }, this.options.retry);

            this._addToBuffer(readMissingBytes);
            return this.readNBytes(numBytes, offset);
        } finally {
            this._lock.release();
        }
    }

    private async _readBytesWithoutRetry(numBytes: number, offset: number) {
        const fd = await fs.open(this.path, this.options.mode);
        try {
            const buffer = Buffer.alloc(numBytes);
            await fd.read(buffer, 0, numBytes, offset);
            return buffer;
        } finally {
            await fd.close();
        }
    }
}
