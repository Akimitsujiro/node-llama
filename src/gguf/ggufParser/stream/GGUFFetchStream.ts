import retry from "async-retry";
import {withLock} from "lifecycle-utils";

import GgufBaseStream, {ALLOCATION_SIZE} from "./GGUFBaseStream.js";

type GGUFFetchStreamOptions = {
    retry: retry.Options
};

export default class GGUFFetchStream extends GgufBaseStream {
    public readonly url: string;
    public readonly options: Partial<GGUFFetchStreamOptions> = {};

    public constructor(url: string, options: Partial<GGUFFetchStreamOptions> = {}) {
        super();
        this.options = options;
        this.url = url;
    }

    public override async readNBytes(numBytes: number, offset = 0): Promise<Buffer> {
        return await withLock(this, "_lock", async function readNBytesWithoutLock(): Promise<Buffer> {
            if (offset + numBytes < this._buffer.length) {
                return this._buffer.subarray(offset, offset + numBytes);
            }

            const fetchMissingBytes = await retry(async () => {
                return await this._fetchBytesWithoutRetry(this._buffer.length, offset + numBytes + ALLOCATION_SIZE);
            }, this.options.retry);

            this._addToBuffer(fetchMissingBytes);
            return await readNBytesWithoutLock.call(this);
        });
    }

    private async _fetchBytesWithoutRetry(start: number, end: number) {
        const response = await fetch(this.url, {
            headers: {
                Range: `bytes=${start}-${end}`,
                accept: "*/*"
            }
        });
        const arrayBuffer = await response.arrayBuffer();
        return Buffer.from(arrayBuffer);
    }
}


