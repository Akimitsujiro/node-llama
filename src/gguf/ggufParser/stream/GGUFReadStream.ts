import fs from "fs/promises";
import retry from "async-retry";
import {withLock} from "lifecycle-utils";
import GgufBaseStream, {ALLOCATION_SIZE} from "./GGUFBaseStream.js";

type GGUFReadStreamOptions = {
    retry?: retry.Options,
    mode: string
};

const DEFAULT_OPTIONS: GGUFReadStreamOptions = {
    mode: "r"
};

export default class GGUFReadStream extends GgufBaseStream {
    public readonly options: GGUFReadStreamOptions;
    public readonly path: string;

    public constructor(path: string, options: Partial<GGUFReadStreamOptions> = {}) {
        super();
        this.path = path;
        this.options = {...DEFAULT_OPTIONS, ...options};
    }

    public override async readNBytes(numBytes: number, offset = 0): Promise<Buffer> {
        return await withLock(this, "_lock", async function readNBytesWithoutLock(): Promise<Buffer> {
            if (offset + numBytes < this._buffer.length) {
                return this._buffer.subarray(offset, offset + numBytes);
            }

            const readMissingBytes = await retry(async () => {
                return await this._readBytesWithoutRetry(numBytes + ALLOCATION_SIZE, this._buffer.length);
            }, this.options.retry);

            this._addToBuffer(readMissingBytes);
            return await readNBytesWithoutLock.call(this);
        });
    }

    private async _readBytesWithoutRetry(numBytes: number, offset: number) {
        const fd = await fs.open(this.path, this.options.mode);
        try {
            const buffer = Buffer.alloc(numBytes);
            await fd.read(buffer, 0, numBytes, offset);
            return buffer;
        } finally {
            await fd.close();
        }
    }
}
