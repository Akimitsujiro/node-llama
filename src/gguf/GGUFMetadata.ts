import retry from "async-retry";
import MetadataNotParsedYetError from "./errors/MetadataNotParsedYetError.js";
import GGUFInsights, {GGUFInsightsOptions} from "./GGUFInsights.js";
import GGUFParser, {GGUFMetadataResponse} from "./ggufParser/GGUFParser.js";
import GGUFFetchStream from "./ggufParser/stream/GGUFFetchStream.js";
import GGUFReadStream from "./ggufParser/stream/GGUFReadStream.js";

export type GGUFMetadataOptions = {
    source?: "network" | "local",
    retry?: retry.Options,
    ignoreKeys?: string[],
    insights?: GGUFInsightsOptions
};

export default class GGUFMetadata {
    protected _metadata?: GGUFMetadataResponse;
    public readonly path: string;
    public readonly options: Partial<GGUFMetadataOptions> = {};

    public get metadata() {
        if (!this._metadata) {
            throw new MetadataNotParsedYetError(this.path);
        }
        return this._metadata;
    }

    public get insights(){
        return new GGUFInsights(this.metadata, this.options.insights);
    }

    public constructor(path: string, options: Partial<GGUFMetadataOptions> = {}) {
        this.options = options;
        this.path = path;
    }

    public async parse() {
        const stream = this._createStream();
        const parser = new GGUFParser(stream, this.options.ignoreKeys);
        return this._metadata = await parser.parseMetadata();
    }

    private _createStream() {
        switch (this.options.source) {
            case "network":
                return new GGUFFetchStream(this.path, {retry: this.options.retry});
            case "local":
            default:
                return new GGUFReadStream(this.path, {retry: this.options.retry});
        }
    }
}
