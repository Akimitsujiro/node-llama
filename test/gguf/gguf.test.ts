import {describe, expect, it} from "vitest";
import GGUFReadStream from "../../src/gguf/gguf-parser/stream/GGUFReadStream.js";
import GGUFParser from "../../src/gguf/gguf-parser/GGUFParser.js";
import GGUFFetchStream from "../../src/gguf/gguf-parser/stream/GGUFFetchStream.js";
import {getModelFile} from "../utils/modelFiles.js";
import GGUFInsights from "../../src/gguf/GGUFInsights.js";
import {getTestLlama} from "../utils/getTestLlama.js";
import {LlamaModel} from "../../src/index.js";
import GGUFMetadata from '../../src/gguf/GGUFMetadata.js';

const remoteGGUFModel = "https://huggingface.co/TheBloke/Falcon-180B-Chat-GGUF/resolve/main/falcon-180b-chat.Q6_K.gguf-split-a?download=true";

describe("GGUF Parser", async () => {
    const modelPath = await getModelFile("functionary-small-v2.2.q4_0.gguf");

    it("Magic should be GGUF local model", async () => {
        const stream = new GGUFReadStream(modelPath);
        const magic = await stream.readNBytes(4);
        const magicText = String.fromCharCode(...magic);

        expect(magicText).toBe("GGUF");
    });

    it("should parse local gguf model", async () => {
        const stream = new GGUFReadStream(modelPath);

        const ggufParser = new GGUFParser(stream);
        const metadata = await ggufParser.parseMetadata();

        expect(metadata).toMatchSnapshot();
    });

    it("Magic should be GGUF remote model", async () => {
        const stream = new GGUFFetchStream(remoteGGUFModel);

        const magic = await stream.readNBytes(4);
        const magicText = String.fromCharCode(...magic);

        expect(magicText).toBe("GGUF");
    });

    it("should parse remote gguf model", async () => {
        const stream = new GGUFFetchStream(remoteGGUFModel);

        const ggufParser = new GGUFParser(stream);
        const metadata = await ggufParser.parseMetadata();

        expect(metadata).toMatchSnapshot();
    }, {timeout: 0});

    it("should calculate GGUF VRAM Usage", async () => {
        const stream = new GGUFReadStream(modelPath);

        const ggufParser = new GGUFParser(stream);
        const metadata = await ggufParser.parseMetadata();

        const ggufInsights = new GGUFInsights(metadata);

        const llama = await getTestLlama();
        new LlamaModel({
            modelPath: modelPath,
            llama
        });

        const usedRam = llama.getVramState().used;

        expect(ggufInsights.VRAMUsage).toMatchInlineSnapshot(`4474643028.666667`);
        expect(usedRam).to.be.gte(3.5 * Math.pow(1024, 3));
        expect(usedRam).to.be.lte(4.5 * Math.pow(1024, 3));
    });

    it("should fetch GGUF metadata", async () => {
        const ggufMetadata = new GGUFMetadata(modelPath);
        await ggufMetadata.parse();

        expect(ggufMetadata.metadata).toMatchSnapshot();
        expect(ggufMetadata.insights.VRAMUsage).toMatchInlineSnapshot(`4474643028.666667`);
    });
});
