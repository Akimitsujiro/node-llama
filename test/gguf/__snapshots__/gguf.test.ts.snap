// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GGUF Parser > should fetch GGUF metadata 1`] = `
{
  "metadata": {
    "general": {
      "architecture": "llama",
      "file_type": "MOSTLY_Q4_0",
      "name": "workspace",
      "quantization_version": 2,
    },
    "llama": {
      "attention": {
        "head_count": 32,
        "head_count_kv": 8,
        "layer_norm_rms_epsilon": 0.000009999999747378752,
      },
      "block_count": 32,
      "context_length": 32768,
      "embedding_length": 4096,
      "feed_forward_length": 14336,
      "rope": {
        "dimension_count": 128,
        "freq_base": 10000,
      },
    },
    "tensorCount": 291,
    "tokenizer": {
      "chat_template": "{% for message in messages %}
{% if message['role'] == 'user' or message['role'] == 'system' %}
{{ '<|from|>' + message['role'] + '
<|recipient|>all
<|content|>' + message['content'] + '
' }}{% elif message['role'] == 'tool' %}
{{ '<|from|>' + message['name'] + '
<|recipient|>all
<|content|>' + message['content'] + '
' }}{% else %}
{% set contain_content='no'%}
{% if message['content'] is not none %}
{{ '<|from|>assistant
<|recipient|>all
<|content|>' + message['content'] }}{% set contain_content='yes'%}
{% endif %}
{% if 'tool_calls' in message and message['tool_calls'] is not none %}
{% for tool_call in message['tool_calls'] %}
{% set prompt='<|from|>assistant
<|recipient|>' + tool_call['function']['name'] + '
<|content|>' + tool_call['function']['arguments'] %}
{% if loop.index == 1 and contain_content == "no" %}
{{ prompt }}{% else %}
{{ '
' + prompt}}{% endif %}
{% endfor %}
{% endif %}
{{ '<|stop|>
' }}{% endif %}
{% endfor %}
{% if add_generation_prompt %}{{ '<|from|>assistant
<|recipient|>' }}{% endif %}",
      "ggml": {
        "bos_token_id": 1,
        "eos_token_id": 2,
        "model": "llama",
        "padding_token_id": 2,
        "unknown_token_id": 0,
      },
    },
    "version": 3,
  },
  "metadataSize": 718762,
}
`;

exports[`GGUF Parser > should parse local gguf model 1`] = `
{
  "metadata": {
    "general": {
      "architecture": "llama",
      "file_type": "MOSTLY_Q4_0",
      "name": "workspace",
      "quantization_version": 2,
    },
    "llama": {
      "attention": {
        "head_count": 32,
        "head_count_kv": 8,
        "layer_norm_rms_epsilon": 0.000009999999747378752,
      },
      "block_count": 32,
      "context_length": 32768,
      "embedding_length": 4096,
      "feed_forward_length": 14336,
      "rope": {
        "dimension_count": 128,
        "freq_base": 10000,
      },
    },
    "tensorCount": 291,
    "tokenizer": {
      "chat_template": "{% for message in messages %}
{% if message['role'] == 'user' or message['role'] == 'system' %}
{{ '<|from|>' + message['role'] + '
<|recipient|>all
<|content|>' + message['content'] + '
' }}{% elif message['role'] == 'tool' %}
{{ '<|from|>' + message['name'] + '
<|recipient|>all
<|content|>' + message['content'] + '
' }}{% else %}
{% set contain_content='no'%}
{% if message['content'] is not none %}
{{ '<|from|>assistant
<|recipient|>all
<|content|>' + message['content'] }}{% set contain_content='yes'%}
{% endif %}
{% if 'tool_calls' in message and message['tool_calls'] is not none %}
{% for tool_call in message['tool_calls'] %}
{% set prompt='<|from|>assistant
<|recipient|>' + tool_call['function']['name'] + '
<|content|>' + tool_call['function']['arguments'] %}
{% if loop.index == 1 and contain_content == "no" %}
{{ prompt }}{% else %}
{{ '
' + prompt}}{% endif %}
{% endfor %}
{% endif %}
{{ '<|stop|>
' }}{% endif %}
{% endfor %}
{% if add_generation_prompt %}{{ '<|from|>assistant
<|recipient|>' }}{% endif %}",
      "ggml": {
        "bos_token_id": 1,
        "eos_token_id": 2,
        "model": "llama",
        "padding_token_id": 2,
        "unknown_token_id": 0,
      },
    },
    "version": 3,
  },
  "metadataSize": 718762,
}
`;

exports[`GGUF Parser > should parse remote gguf model 1`] = `
{
  "metadata": {
    "falcon": {
      "attention": {
        "head_count": 232,
        "head_count_kv": 8,
        "layer_norm_epsilon": 0.000009999999747378752,
      },
      "block_count": 80,
      "context_length": 2048,
      "embedding_length": 14848,
      "feed_forward_length": 59392,
      "tensor_data_layout": "jploski",
    },
    "general": {
      "architecture": "falcon",
      "file_type": "MOSTLY_Q6_K",
      "name": "Falcon",
      "quantization_version": 2,
    },
    "tensorCount": 644,
    "tokenizer": {
      "ggml": {
        "eos_token_id": 11,
        "model": "gpt2",
      },
    },
    "version": 2,
  },
  "metadataSize": 2547826,
}
`;
